Formula/ghc.rb

```diff
@@ -40,34 +41,87 @@ class Ghc < Formula
     end
   end
 
+  patch do
+    url "https://gist.githubusercontent.com/hjelmn/49f70df40da0399006031d854f73473e/raw/601bed7124f0fb9787d94806eab61645b85dab0a/ghc-8.10.3.patch"
+    sha256 "5c4747103802e717ea20af05898387058b07bf0a73d38bb14c6e1ea2173baba6"
+  end
+
   def install
     ENV["CC"] = ENV.cc
     ENV["LD"] = "ld"
     ENV["PYTHON"] = Formula["python@3.9"].opt_bin/"python3"
 
-    # Build a static gmp rather than in-tree gmp, otherwise all ghc-compiled
-    # executables link to Homebrew's GMP.
-    gmp = libexec/"integer-gmp"
-
-    # GMP *does not* use PIC by default without shared libs so --with-pic
-    # is mandatory or else you'll get "illegal text relocs" errors.
-    resource("gmp").stage do
-      system "./configure", "--prefix=#{gmp}", "--with-pic", "--disable-shared",
-                            "--build=#{Hardware.oldest_cpu}-apple-darwin#{OS.kernel_version.major}"
-      system "make"
-      system "make", "install"
-    end
-
-    args = ["--with-gmp-includes=#{gmp}/include",
-            "--with-gmp-libraries=#{gmp}/lib"]
+    args = []
 
     resource("binary").stage do
       binary = buildpath/"binary"
 
       system "./configure", "--prefix=#{binary}", *args
       ENV.deparallelize { system "make", "install" }
 
-      ENV.prepend_path "PATH", binary/"bin"
+      ENV["GHC"] = binary/"bin/ghc"
+    end
+
+    cpu = if Hardware::CPU.arm?
+      "arm64"
+    else
+      Hardware.oldest_cpu
+    end
+
+    # Replace in-tree gmp with 6.2.1 which supports arm64.
+    rm "libraries/integer-gmp/gmp/gmp-tarballs/gmp-6.1.2-nodoc.tar.bz2", force: true
+    cp resource("gmp").cached_download, "libraries/integer-gmp/gmp/gmp-tarballs/gmp-6.2.1-nodoc.tar.bz2"
+
+    # There is no native codegen for ARM64 yet.
+    if Hardware::CPU.arm?
+      args << "--enable-unregisterised"
```

Comment:
You should not need to build an unregistered build. The llvm backend should just work fine on arm64. We used to have some bare support for iphone from 8.4 onwards already even though it wasn't fully sound.

Reply:
For the LLVM build, you'll need to depend on a recent llvm (and can pretty safely ignore the "not supported LLVM version warnings", the llvm IR hasn't changed to the point where it breaks as far as I can tell). Make sure to set LLC and OPT to the respective tool paths.

Reply:
Are there any special requirements from the LLVM backend, or can ghc use Homebrew's LLVM? https://github.com/Homebrew/homebrew-core/blob/master/Formula/llvm.rb

Reply:
Aside from having opt and llc and supporting the necessary targets, no there are no special requirements.

ghc will produce textual llvm Ir, feed that into opt, and then into llc. Then GHC will perform some sadly necessary assembly fixups on assembly file and feed that into `as` to produce the object file.